<?xml version="1.0" encoding="UTF-8" ?>

<launch>
  <master auto="start"/>
  <param name="/use_sim_time" value="true"/>

  <!--Pionner 3dx -->
  <arg name="model" default="$(find p3dx_description)/urdf/pioneer3dx.xml"/>
  <param name="robot_description" textfile="$(arg model)" /> 

  <!--Stage World
  <include file="$(find rc_simul_worlds)/launch/2pioneers_stage_MR_SLAM.launch" /> -->
  <group ns="robot_0">
    <arg name="prefix" value="robot_0"/>
    <arg name="init_pose_x" value="2.0"/>
    <arg name="init_pose_y" value="2.0"/>
    <arg name="init_pose_z" value="0.0"/>
    <arg name="init_pose_yaw" value="0.0"/>

    <include file="$(find p3dx_description)/launch/p3dx_description.launch">
      <arg name="tf_prefix" value="$(arg prefix)" />
    </include>

    <node name="robot_laser_grid_mapping" pkg="robot_laser_grid_mapping" type="grid_mapping.py" output="screen">
        <param name="sensor_model_p_occ"        value="0.99"/>
        <param name="sensor_model_p_free"       value="0.01"/> 
        <param name="sensor_model_p_prior"      value="0.5"/> 
        <param name="robot_frame"               value="$(arg prefix)/base_link"/> 
        <param name="map_frame"                 value="$(arg prefix)/map"/> 
        <param name="map_center_x"              value="-1.0"/> 
        <param name="map_center_y"              value="-1.0"/> 
        <param name="map_size_x"                value="32.0"/> 
        <param name="map_size_y"                value="12.0"/> 
        <param name="map_resolution"            value="0.1"/> 
        <param name="map_publish_freq"          value="0.5"/> <!-- IMPORTANTE 1/map_publish_freq (segundos)-->
        <param name="update_movement"           value="0.1"/> <!-- NAO ESTÁ A SER USADO -->
    </node>

    <node pkg="tf" type="static_transform_publisher" name="link1_broadcaster_$(arg prefix)" args="0 0 0 0 0 0 1 $(arg prefix)/base_link $(arg prefix)/laser 100" />

    <node name="robot_laser_simulator" pkg="robot_laser_simulator" type="simulator.py" output="screen">
        <param name="map_file" value="$(find robot_laser_simulator)/launch/map.png"/>
        <param name="map_resolution" value="0.05"/> <!-- meters per cell -->
        <param name="time_resolution" value="0.1"/> <!-- time diff between simulator ticks in seconds -->
        <param name="laser_min_angle" value="-135"/> <!-- laser start angle in degrees -->
        <param name="laser_max_angle" value="135"/> <!-- laser end angle in degrees -->
        <param name="laser_resolution" value="1"/> <!-- angle between two laser rays in degrees -->
        <param name="laser_noise_mu" value="0.1"/> <!-- additive normal distribution noise -->
        <param name="laser_noise_sigma" value="0.02"/> <!-- additive normal distribution noise -->
        <param name="laser_max_dist" value="7.0"/> <!-- rays exceeding this value will be converted no NaN -->
        <param name="robot_pos_x" value="2.0"/> <!-- start point -->
        <param name="robot_pos_y" value="2.0"/> <!-- start point -->
        <param name="robot_pos_theta" value="0"/> <!-- start point -->
        <param name="odom_frame" value=""/> <!-- leave empty for disabling odom publishing -->
        <param name="robot_frame" value="$(arg prefix)/base_link"/> <!-- also modify static_transform_publisher -->
        <param name="laser_frame" value="$(arg prefix)/laser"/> <!-- also modify static_transform_publisher -->
    </node>

    <node pkg="tf" type="static_transform_publisher" name="static_tf_$(arg prefix)_to_map" args="0.0 0.0 0.0 0.0 0.0 0.0 map $(arg prefix)/map 10"  />
    <node pkg="tf" type="static_transform_publisher" name="static_tf_$(arg prefix)_to_base_link" args="0.0 0.0 0.0 0.0 0.0 0.0 $(arg prefix)/map $(arg prefix)/base_link 10"  />

    <node pkg="move_base" type="move_base" respawn="false" name="move_base_node" output="screen">
      <!--remap from="$(arg namespace)/scan" to="scan" /-->
    
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/costmap_common_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/costmap_common_params.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/move_base_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/local_planner_params.yaml" command="load" />

      <param name="TrajectoryPlannerROS/yaw_goal_tolerance" value="3.14" /> 
        
      <param name="global_costmap/robot_base_frame" value="$(arg prefix)/base_link"/>   
      <param name="local_costmap/global_frame" value="$(arg prefix)/map"/>
      <param name="TrajectoryPlannerROS/global_frame" value="$(arg prefix)/map"/>
      <param name="global_costmap/obstacles/scan/sensor_frame" value="$(arg prefix)/base_laser_link"/>
      <param name="global_costmap/obstacles/scan/topic" value="$(arg prefix)/scan"/>
      <param name="local_costmap/robot_base_frame" value="$(arg prefix)/base_link"/> 
      <param name="local_costmap/obstacles/scan/sensor_frame" value="$(arg prefix)/base_laser_link"/>
      <param name="local_costmap/obstacles/scan/topic" value="$(arg prefix)/scan"/>
    </node>
  </group>

  <group ns="robot_1">
    <arg name="prefix" value="robot_1"/>
    <arg name="init_pose_x" value="3.0"/>
    <arg name="init_pose_y" value="3.0"/>
    <arg name="init_pose_z" value="0.0"/>
    <arg name="init_pose_yaw" value="0.0"/>

    <include file="$(find p3dx_description)/launch/p3dx_description.launch">
      <arg name="tf_prefix" value="$(arg prefix)" />
    </include>

    <node name="robot_laser_grid_mapping" pkg="robot_laser_grid_mapping" type="grid_mapping.py" output="screen">
        <param name="sensor_model_p_occ"        value="0.99"/>
        <param name="sensor_model_p_free"       value="0.01"/> 
        <param name="sensor_model_p_prior"      value="0.5"/> 
        <param name="robot_frame"               value="$(arg prefix)/base_link"/> 
        <param name="map_frame"                 value="$(arg prefix)/map"/> 
        <param name="map_center_x"              value="-1.0"/> 
        <param name="map_center_y"              value="-1.0"/> 
        <param name="map_size_x"                value="32.0"/> 
        <param name="map_size_y"                value="12.0"/> 
        <param name="map_resolution"            value="0.1"/> 
        <param name="map_publish_freq"          value="0.5"/> <!-- IMPORTANTE 1/map_publish_freq (segundos)-->
        <param name="update_movement"           value="0.1"/> <!-- NAO ESTÁ A SER USADO -->
    </node>

    <node pkg="tf" type="static_transform_publisher" name="link1_broadcaster_$(arg prefix)" args="0 0 0 0 0 0 1 $(arg prefix)/base_link $(arg prefix)/laser 100" />

    <node pkg="tf" type="static_transform_publisher" name="static_tf_$(arg prefix)_to_map" args="0.0 0.0 0.0 0.0 0.0 0.0 map $(arg prefix)/map 10"  />
    <node pkg="tf" type="static_transform_publisher" name="static_tf_$(arg prefix)_to_base_link" args="0.0 0.0 0.0 0.0 0.0 0.0 $(arg prefix)/map $(arg prefix)/base_link 10"  />

    <node pkg="move_base" type="move_base" respawn="false" name="move_base_node" output="screen">
      <!--remap from="$(arg namespace)/scan" to="scan" /-->
    
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/costmap_common_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/costmap_common_params.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/move_base_params.yaml" command="load" />
      <rosparam file="$(find rc_simul_worlds)/params/pioneer/local_planner_params.yaml" command="load" />

      <param name="TrajectoryPlannerROS/yaw_goal_tolerance" value="3.14" /> 
        
      <param name="global_costmap/robot_base_frame" value="$(arg prefix)/base_link"/>   
      <param name="local_costmap/global_frame" value="$(arg prefix)/map"/>
      <param name="TrajectoryPlannerROS/global_frame" value="$(arg prefix)/map"/>
      <param name="global_costmap/obstacles/scan/sensor_frame" value="$(arg prefix)/base_laser_link"/>
      <param name="global_costmap/obstacles/scan/topic" value="$(arg prefix)/scan"/>
      <param name="local_costmap/robot_base_frame" value="$(arg prefix)/base_link"/> 
      <param name="local_costmap/obstacles/scan/sensor_frame" value="$(arg prefix)/base_laser_link"/>
      <param name="local_costmap/obstacles/scan/topic" value="$(arg prefix)/scan"/>
    </node>
  </group>

    <!--node pkg="explore_lite" type="explore" respawn="false" name="explore" output="screen">
      <param name="robot_base_frame" value="base_link"/>
      <param name="costmap_topic" value="/map"/>
      <param name="costmap_updates_topic" value="map_updates"/>
      <param name="visualize" value="false"/>
      <param name="planner_frequency" value="0.33"/>
      <param name="progress_timeout" value="30.0"/>
      <param name="potential_scale" value="3.0"/>
      <param name="orientation_scale" value="0.0"/>
      <param name="gain_scale" value="1.0"/>
      <param name="transform_tolerance" value="0.3"/>
      <param name="min_frontier_size" value="0.5"/>
    </node-->

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find first_robot_navigation)/Rviz/Two.rviz"/>

</launch>


